src/vec3.rs

```diff
@@ -1,324 +1,219 @@
-            z: self * rhs.z,
-        }
-    }
-}
+    pub fn random_in_unit_disk() -> Vec3 {
+        let mut rng = rand::thread_rng();
 
-impl ops::Mul<&Vec3> for f64 {
-    type Output = Vec3;
-    fn mul(self, rhs: &Vec3) -> Self::Output {
-        Vec3 {
-            x: self * rhs.x,
-            y: self * rhs.y,
-            z: self * rhs.z,
+        // TODO: this seems suboptimal, is it used often?
+        loop {
+            let p = Vec3::new(rng.gen_range(-1.0..1.0), rng.gen_range(-1.0..1.0), 0.0);
+            if p.length_squared() < 1.0 {
+                return p;
+            }
         }
     }
-}
 
-impl ops::Div<f64> for Vec3 {
-    type Output = Vec3;
     #[inline]
-    fn div(self, rhs: f64) -> Self::Output {
-        Vec3 {
-            x: self.x / rhs,
-            y: self.y / rhs,
-            z: self.z / rhs,
-        }
+    pub fn random_unit_vector(rng: &mut impl rand::Rng) -> Vec3 {
+        Vec3::random_in_unit_sphere(rng).normalize()
     }
-}
 
-impl ops::Div<f64> for &Vec3 {
-    type Output = Vec3;
     #[inline]
-    fn div(self, rhs: f64) -> Self::Output {
-        Vec3 {
-            x: self.x / rhs,
-            y: self.y / rhs,
-            z: self.z / rhs,
-        }
+    pub fn near_zero(&self) -> bool {
+        // TODO: check if is_normal() is not enough
+        const S: f64 = 1.0e-8;
+        self.x.abs() < S && self.y.abs() < S && self.z.abs() < S
     }
-}
 
-impl ops::Neg for Vec3 {
-    type Output = Vec3;
     #[inline]
-    fn neg(self) -> Self::Output {
-        Vec3 {
-            x: -self.x,
-            y: -self.y,
-            z: -self.z,
-        }
+    pub fn reflect(self, normal: Vec3) -> Vec3 {
+        self - 2.0 * self.dot(normal) * normal
     }
-}
 
-impl ops::Neg for &Vec3 {
-    type Output = Vec3;
     #[inline]
-    fn neg(self) -> Self::Output {
-        Vec3 {
-            x: -self.x,
-            y: -self.y,
-            z: -self.z,
-        }
-    }
-}
-
-impl ops::AddAssign<&Vec3> for Vec3 {
-    fn add_assign(&mut self, rhs: &Vec3) {
-        self.x += rhs.x;
-        self.y += rhs.y;
-        self.z += rhs.z;
-    }
-}
-
-impl ops::AddAssign<Vec3> for Vec3 {
-    fn add_assign(&mut self, rhs: Vec3) {
-        self.x += rhs.x;
-        self.y += rhs.y;
-        self.z += rhs.z;
-    }
-}
-
-impl ops::MulAssign<f64> for Vec3 {
-    fn mul_assign(&mut self, rhs: f64) {
-        self.x *= rhs;
-        self.y *= rhs;
-        self.z *= rhs;
-    }
-}
-
-impl ops::DivAssign<f64> for Vec3 {
-    fn div_assign(&mut self, rhs: f64) {
-        self.x /= rhs;
-        self.y /= rhs;
-        self.z /= rhs;
+    pub fn refract(self, n: Vec3, etai_over_etat: f64) -> Vec3 {
+        let uv = self;
+        let cos_theta = (-uv.dot(n)).min(1.0);
+        let r_out_perp = etai_over_etat * (uv + cos_theta * n);
+        let r_out_parallel = -(1.0 - r_out_perp.length_squared()).sqrt() * n;
+        r_out_perp + r_out_parallel
     }
 }
 
 impl fmt::Display for Vec3 {
-    #[inline]
```

Comment:
Why remove the inline specifier?

Reply:
Ah, I forgot about that.
None of these `#[inline]`s actually do anything.
They only tell the compiler to inline bewteen separate crates when compiling with LTO. We neither compiler with LTO, nor do we use these function in separate crates. I think we should either replace them with `#[inline(always)]` where benchmarking shows it makes sense or remove them and let LLVM  figure out what to inline.

Reply:
That is probably something that can also confuse newcomers from C++.